### PR TITLE
Fixed groupby error with nullable types

### LINKED ISSUE #54206: BUG: Grouped rank incorrect behaviour with nullable types
### Pandas version checks

- [X] I have checked that this issue has not already been reported.

- [X] I have confirmed this bug exists on the [latest version](https://pandas.pydata.org/docs/whatsnew/index.html) of pandas.

- [X] I have confirmed this bug exists on the [main branch](https://pandas.pydata.org/docs/dev/getting_started/install.html#installing-the-development-version-of-pandas) of pandas.

### Reproducible Example

```python
import pandas as pd

# numpy dtype (None -> np.NaN)
# This will print the following:
# 
# 0    1.0
# Name: x, dtype: float64
#
# This is the expected behaviour

df_np = pd.DataFrame({ "x": [None] }, dtype="float64")
print(df_np.groupby("x", dropna=False)["x"].rank(method="min", na_option="bottom"))


# pandas nullable dtype (None -> pd.NA)
# This will print the following:
#
# 0    <NA>
# Name: x, dtype: Float64
#
# This is NOT the expected behaviour, because the rank is NA instead of 1

df_ext = pd.DataFrame({ "x": [None] }, dtype="Float64")
print(df_ext.groupby("x", dropna=False)["x"].rank(method="min", na_option="bottom"))
```

### Issue Description

When using the rank function on grouped data frames, the result is different when using nullable datatypes compared to numpy datatypes.
This is only an issue when passing `na_option="bottom"` or `na_option="top"`.

### Expected Behavior

The result shouldn't depend on the datatype that got used.

### Installed Versions

<details>

INSTALLED VERSIONS
------------------
commit           : 0f437949513225922d851e9581723d82120684a6
python           : 3.9.6.final.0
python-bits      : 64
OS               : Darwin
OS-release       : 22.5.0
Version          : Darwin Kernel Version 22.5.0: Thu Jun  8 22:22:20 PDT 2023; root:xnu-8796.121.3~7/RELEASE_ARM64_T6000
machine          : arm64
processor        : arm
byteorder        : little
LC_ALL           : None
LANG             : en_US.UTF-8
LOCALE           : en_US.UTF-8

pandas           : 2.0.3
numpy            : 1.25.1
pytz             : 2023.3
dateutil         : 2.8.2
setuptools       : 68.0.0
pip              : 23.1.2
Cython           : None
pytest           : 7.4.0
hypothesis       : None
sphinx           : 7.0.1
blosc            : None
feather          : None
xlsxwriter       : None
lxml.etree       : None
html5lib         : None
pymysql          : None
psycopg2         : 2.9.6
jinja2           : 3.1.2
IPython          : 8.14.0
pandas_datareader: None
bs4              : 4.12.2
bottleneck       : None
brotli           : None
fastparquet      : None
fsspec           : None
gcsfs            : None
matplotlib       : None
numba            : None
numexpr          : None
odfpy            : None
openpyxl         : None
pandas_gbq       : None
pyarrow          : None
pyreadstat       : None
pyxlsb           : None
s3fs             : None
scipy            : None
snappy           : None
sqlalchemy       : 2.0.19
tables           : None
tabulate         : None
xarray           : None
xlrd             : None
zstandard        : None
tzdata           : 2023.3
qtpy             : None
pyqt5            : None
</details>

### PR DESCRIPTION
**Pull request by KladeRe**
- [x] closes #54206 
- [x] [Tests added and passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#writing-tests) if fixing a bug or adding a new feature
- [x] All [code checks passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#pre-commit).
- [x] Added [type annotations](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#type-hints) to new arguments/methods/functions.
- [x] Added an entry in the latest `doc/source/whatsnew/vX.X.X.rst` file if fixing a bug or adding a new feature.

### CODE REVIEW

File: pandas/tests/groupby/test_rank.py [hunk 1]
```diff
@@ -710,3 +710,9 @@ def test_rank_categorical():
 
     expected = df.astype(object).groupby("col1").rank()
     tm.assert_frame_equal(res, expected)
+
+def test_groupy_op_with_nullables():
```
Comment (el33tneel) [inline pandas/tests/groupby/test_rank.py:714]: Minor: typo in method name: `test_groupy_op_with_nullables`-> `test_groupby_op_with_nullables`

File: doc/source/whatsnew/v2.1.0.rst [hunk 2]
```diff
@@ -183,10 +183,13 @@ Notable bug fixes
 
 These are bug fixes that might have notable behavior changes.
 
-.. _whatsnew_210.notable_bug_fixes.notable_bug_fix1:
+.. _fixed_groupby_error_with_nullable_types:
 
-notable_bug_fix1
+groupby_nullable_types_error
```
Comment (rhshadrach) [inline doc/source/whatsnew/v2.1.0.rst:188]: I don't think this qualifies for a notable bug fix. Instead, could you add to the GroupBy subsection of the bugfixes down below this.

File: pandas/core/arrays/masked.py [hunk 3]
```diff
@@ -1440,10 +1440,8 @@ def _groupby_op(
 
         # libgroupby functions are responsible for NOT altering mask
         mask = self._mask
-        if op.kind != "aggregate":
-            result_mask = mask.copy()
-        else:
-            result_mask = np.zeros(ngroups, dtype=bool)
+
+        result_mask = np.zeros(ngroups, dtype=bool)
```
Comment (rhshadrach) [inline pandas/core/arrays/masked.py:1444]: This does not implement the change I suggested in the linked issue, and I believe is causing tests to fail.

**Review by rhshadrach — CHANGES_REQUESTED** [review]

**Comment by mroeschke** [comment]
Thanks for the pull request, but it appears to have gone stale. If interested in continuing, please merge in the main branch, address any review comments and/or failing tests, and we can reopen.

**Pull request closed without merging**